x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java

```diff
@@ -403,6 +403,47 @@ public static FunctionDefinition def(Class<? extends Function> function, Functio
         T build(Source source, Expression src, Expression exp1, Expression exp2, Expression exp3);
     }
 
+
+    // Translation table for error messaging in the following function
+    private static final String[] NUM_NAMES = {
+            "zero",
+            "one",
+            "two",
+            "three",
+            "four",
+            "five",
+    };
+
+    @SuppressWarnings("overloads")  // These are ambiguous if you aren't using ctor references but we always do
+    public static <T extends Function> FunctionDefinition def(Class<T> function,
+                                                              FiveParametersFunctionBuilder<T> ctorRef,
+                                                              int numOptionalParams, String... names) {
+        FunctionBuilder builder = (source, children, distinct, cfg) -> {
+            final int NUM_TOTAL_PARAMS = 5;
+            boolean hasOptionalParams = OptionalArgument.class.isAssignableFrom(function);
+            if (hasOptionalParams && (children.size() > NUM_TOTAL_PARAMS || children.size() < NUM_TOTAL_PARAMS - numOptionalParams)) {
+                throw new QlIllegalArgumentException("expects between " + NUM_NAMES[NUM_TOTAL_PARAMS - numOptionalParams]
+                        + " and " + NUM_NAMES[NUM_TOTAL_PARAMS] + " arguments");
+            } else if (!hasOptionalParams && children.size() != NUM_TOTAL_PARAMS) {
+                throw new QlIllegalArgumentException("expects exactly " + NUM_NAMES[NUM_TOTAL_PARAMS] + " arguments");
```

Comment:
I understand this wants to be a generic approach for dealing with a variable number of optional parameters, but:
- this solution needs to be generic throughout this class: [here](https://github.com/elastic/elasticsearch/blob/be0b67cdf5022f38ec3610a68bbb91af8b10b8b3/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java#L318) it's another method accepting optional parameters.
- I don't understand the purpose of this method... it tries to be generic, but `NUM_TOTAL_PARAMS` is always 5. So, this method is for functions that can have between 0 and 5 optional parameters? Apart from `between` is there any other such function?

Reply:
The approach by the `here` link in you comment allows only one optional parameter. 
Here I created a kind of more generic implementation allowing to specify the number of optional parameters.
In the case of `between` function we need to support up to two optional parameters.

Reply:
Let's revisit this based on whether `between` actually needs case sensitivity or not - if it doesn't it will simplify things.
Same applies for greedy.

/cc @rw-access